crates/rust-analyzer/src/handlers/request.rs

```diff
@@ -1410,17 +1410,71 @@ pub(crate) fn handle_inlay_hints(
         snap.analysis
             .inlay_hints(&inlay_hints_config, file_id, Some(range))?
             .into_iter()
-            .map(|it| to_proto::inlay_hint(&snap, &line_index, it))
+            .map(|it| {
+                to_proto::inlay_hint(
+                    &snap,
+                    &inlay_hints_config.fields_to_resolve,
+                    &line_index,
+                    file_id,
+                    it,
+                )
+            })
             .collect::<Cancellable<Vec<_>>>()?,
     ))
 }
 
 pub(crate) fn handle_inlay_hints_resolve(
-    _snap: GlobalStateSnapshot,
-    hint: InlayHint,
+    snap: GlobalStateSnapshot,
+    mut original_hint: InlayHint,
 ) -> anyhow::Result<InlayHint> {
     let _p = profile::span("handle_inlay_hints_resolve");
-    Ok(hint)
+
+    let data = match original_hint.data.take() {
+        Some(it) => it,
+        None => return Ok(original_hint),
+    };
+
+    let resolve_data: lsp_ext::InlayHintResolveData = serde_json::from_value(data)?;
+    let file_id = FileId(resolve_data.file_id);
+    let line_index = snap.file_line_index(file_id)?;
+    let range = from_proto::text_range(
+        &line_index,
+        lsp_types::Range { start: original_hint.position, end: original_hint.position },
+    )?;
+    let range_start = range.start();
+    let range_end = range.end();
+    let large_range = TextRange::new(
+        range_start.checked_sub(1.into()).unwrap_or(range_start),
+        range_end.checked_add(1.into()).unwrap_or(range_end),
+    );
+    let mut forced_resolve_inlay_hints_config = snap.config.inlay_hints();
+    forced_resolve_inlay_hints_config.fields_to_resolve = InlayFieldsToResolve::empty();
+    let resolve_hints = snap.analysis.inlay_hints(
+        &forced_resolve_inlay_hints_config,
+        file_id,
+        Some(large_range),
```

Comment:
This won't necessarily suffice as we have some hints that are derived from parent nodes (like lifetime hints) where the "large range" might not capture the node still (unsure if we can actually hit this right now, but just raisin it to be aware). This is kind of a flaw in how we calculate the inlay hints with a supplied `range_limit` in general though

Reply:
I could save the query range inside `InlayHintResolveData` and re-query it entirely, but that sounds wasteful.
Do we have a notion of the "source range" in such hints? 
Otherwise, not sure what I can do about it, except for leaving as is for now.

Reply:
Ye I am not sure either, this is pretty much an unsolved problem (as it already exists with the ranged semantic tokens request which might miss out on hints as well in this case).

In fact I can already think of a case where this breaks, lifetime hints, `fn f(_: &b)` will have hints as `fn f<'0>(_: &'0 b)`, so hovering the lifetime hint in the reference will look at `&` with the large range being ` &(` so we will look at most at the parameter list as the node, missing out on the function which is what kicks off lifetime elision hints. It's not too bad right now, because those hints have no resolvable data anyways. But we need to keep this in mind (that is open an issue about this at least). Alternatively for these known problem cases maybe we can just be eager and not resolve them lazily to somewhat get around the issue for the time being.

Reply:
> Alternatively for these known problem cases maybe we can just be eager and not resolve them lazily to somewhat get around the issue for the time being.

The problem with this would be to understand which could be resolved lazily and which could not.
Ironically, best heuristics I could think of is exactly the position difference (we report the hint for one position, but inferred it from another position on the document).
I have not found a good way to know that yet, so I might lean towards creating a follow-up issue, most probably we can mention it as a comment in https://github.com/rust-lang/rust-analyzer/issues/13962 since this PR does not close it.

Reply:
:+1:

Reply:
I have summarized the PR in the comment: https://github.com/rust-lang/rust-analyzer/issues/13962#issuecomment-1703917142 so feels like this last one is resolved too?